crates/red_knot_python_semantic/src/types.rs

```diff
@@ -822,6 +822,14 @@ impl<'db> CallOutcome<'db> {
                             called_ty.display(db)
                         ),
                     ),
+                    _ if not_callable.len() == outcomes.len() => builder.add_diagnostic(
+                        node,
+                        "call-non-callable",
+                        format_args!(
+                            "No element of union type '{}' is callable.",
```

Comment:
I expect you'll call me pedantic here, but this feels like it confuses the _type_ being (not) callable with _instances of the type_ being (not) callable. What about something like this?

```suggestion
                            "No possible instance of type '{}' is callable.",
```

Reply:
After considering, I think what makes most sense is to just change this message to say `Object of type 'foo | bar' is not callable.` the same as we'd do for a non-union non-callable type. There's really no difference.

I think your objection might also apply to the existing `Union element '{}' of type '{}' is not callable.` and `Union elements {} of type '{}' are not callable.` messages, but since those aren't touched in this PR and it's not clear to me how to improve them, I'll let you propose a separate change to them if you want.

I feel like this is a mis-interpretation of what a "type" means, since "the type `Foo`" inherently means instances of Foo, so "the type `Foo` is not callable" already means "instances of Foo are not callable" -- as distinct from saying "the type `Type[Foo]` (or `type`, or `Literal[Foo]`) is not callable". So I'm not totally convinced we want to make our error messages more verbose to extra-clarify this. But if there are ways to word them that aren't super verbose and make this clearer, I don't have a problem with that.

Reply:
No, "object of type X is not callable" feels correct to me whereas "type X is not callable" does not feel correct. It is the object that has that type — the instance of the type — that is not callable. The not-callable type is not a subtype of the Callable protocol, but the type itself is actually callable, of course (in the vast majority of instances); that's how you create instances of the type, of course: by calling it. Or, well, perhaps you could argue that in fact you call the _class_ rather than the _type_ to construct instances of the not-callable type — but I suspect that's a subtlety that'll be lost on most of our users.

Reply:
it seems like neither pyright nor mypy even consider this problem, because they simply issue a separate diagnostic for each element in the union that isn't a subtype of `Callable`, rather than trying to consolidate them into a single diagnostic like we do 😄

- https://mypy-play.net/?mypy=latest&python=3.12&gist=814f33f5253008f2698adc4913f4ac71
- https://pyright-play.net/?pythonVersion=3.13&strict=true&code=CYUwZgBGAUAeBcECWA7ALhAPhAzmgTgJTwCwAUBJRLNIeUA

Reply:
I tweaked the other combined union messages so that all the messages are more consistent in form.

Reply:
Thank you!